File: framework/examples/talker.cpp
/******************************************************************************
 * Copyright 2018 The Apollo Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *****************************************************************************/

#include <thread>

#include "cybertron/cybertron.h"
#include "cybertron/proto/chatter.pb.h"
#include "cybertron/time/rate.h"
#include "cybertron/time/time.h"

using apollo::cybertron::Rate;
using apollo::cybertron::Time;
using apollo::cybertron::proto::Chatter;

int main(int argc, char *argv[]) {
  // init cybertron framework
  apollo::cybertron::Init(argv[0]);
  // create talker node
  std::shared_ptr<apollo::cybertron::Node> talker_node(
      apollo::cybertron::CreateNode("talker"));
  // create talker
  auto talker = talker_node->CreateWriter<Chatter>("channel/chatter");
  Rate rate(1.0);
  while (apollo::cybertron::OK()) {
    static uint64_t seq = 0;
    auto msg = std::make_shared<apollo::cybertron::proto::Chatter>();
    msg->set_timestamp(Time::Now().ToNanosecond());
    msg->set_lidar_timestamp(Time::Now().ToNanosecond());
    msg->set_seq(seq++);
    msg->set_content("Hello, apollo!");
    talker->Write(msg);
    AINFO << "talker sent a message!";
    rate.Sleep();
  }
  return 0;
}

File: framework/python/examples/talker.py
import time
import sys

sys.path.append("../")
from cybertron import node
from cybertron import init
from proto import chatter_pb2

def test_talker_class():
    msg = chatter_pb2.Chatter()
    msg.content = "talker:send Alex!"
    msg.seq = 0
    msg.timestamp = 0
    msg.lidar_timestamp = 0

    test_node = node.Node("node_name1")
    g_count = 1

    w = test_node.create_writer("channel/chatter", chatter_pb2.Chatter.DESCRIPTOR.full_name)
    while not test_node.is_shutdown():  
        time.sleep(1)
        g_count = g_count + 1
        msg.seq = g_count
        print("="*80)
        print("write msg -> %s" % msg)
        w.write(msg)

def main():
    init.init()
    test_talker_class()
    init.shutdown()

if __name__ == '__main__':
  main()